How many late days are you using for this assignment? One day

In [4]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [5]:
# Load the given datasets with correct filenames
df_num = pd.read_csv('chronic_kidney_disease_numerical.csv')
df_cat = pd.read_csv('chronic_kidney_disease_categorical.csv')

# Print the data
print(df_num)
print("\n\n")
print(df_cat)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
This dataset is a collection of medical records that can be used to identify whether a person has chronic kidney disease (CKD). Each record includes various health indicators such as age, blood pressure, urine test results (like sugar, specific gravity, albumin levels), and blood test results (like blood glucose, urea, creatinine). It also includes several yes/no indicators for conditions like hypertension, diabetes, and anemia. The goal is to use these measurements to predict whether the patient has chronic kidney disease (the “class” variable). Essentially, it’s a diagnostic dataset that captures a combination of numerical lab values and categorical health conditions to help in detecting CKD.

Features: All columns except for class (i.e., age, bp, sg, al, su, rbc, pc, pcc, ba, bgr, bu, sc, sod, pot, hemo, pcv, wbcc, rbcc, htn, dm, cad, appet, pe, ane).
Label/Target: The class column, which can be “ckd” or “notckd.”

This dataset contains both categorical and numerical features.

Numerical features: Examples include age, bp (blood pressure), bgr (blood glucose random), bu (blood urea), sc (serum creatinine), etc.
Categorical features: Examples include sg (specific gravity), rbc (normal/abnormal), htn (yes/no), etc.

Categorical (nominal/binary) examples:
sg (specific gravity) – takes values like 1.005, 1.010, etc.
rbc (red blood cells) – “normal” or “abnormal.”
htn (hypertension) – “yes” or “no.”

Numerical examples:
age – measured in years.
bp (blood pressure) – measured in mm/Hg.
bgr (blood glucose random) – measured in mg/dl.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [11]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_duplicates = df_num.duplicated().sum()
print(f"Numerical dataset")
print(f"Number of duplicate rows: {num_duplicates}")

# Drop these duplicate rows
if num_duplicates > 0:
    print("\nDuplicate rows in numerical dataset:")
    print(df_num[df_num.duplicated(keep='first')])
    
    # Drop duplicates from numerical dataset
    df_num_clean = df_num.drop_duplicates(keep='first')
    print(f"\nShape before dropping duplicates: {df_num.shape}")
    print(f"Shape after dropping duplicates: {df_num_clean.shape}")

print("\n" + "="*50 + "\n")

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_duplicates = df_cat.duplicated().sum()
print("Categorical Dataset:")
print(f"Number of duplicate rows: {cat_duplicates}")

if cat_duplicates > 0:
    print("\nDuplicate rows in categorical dataset:")
    print(df_cat[df_cat.duplicated(keep='first')])
    
    # Drop duplicates from categorical dataset
    df_cat_clean = df_cat.drop_duplicates(keep='first')
    print(f"\nShape before dropping duplicates: {df_cat.shape}")
    print(f"Shape after dropping duplicates: {df_cat_clean.shape}")


Numerical dataset
Number of duplicate rows: 30

Duplicate rows in numerical dataset:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
107     872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
149     721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
171     240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
173     441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
185     261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
194     769512  38.0   60.0   91.0   36.0  0.7  135.0  3.7  14.0  46.0   
213     354223  64.0   70.0   97.0   27.0  0.7  145.0  4.8  13.8  49.0   
244     302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
245     431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
248     687021  52.0   80.0   99.0   25.0  0.8  135.0  3.7  15.0  52.0   
249     843449  30.0   80.0  131.0   38.0  1.0  147.0  3.8  14.1  45.0   
265     628271  65.0   90.0

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: I do not believe that all unique ids are present in both datasets since both datasets do not have the same number of rows. And a row represents the data on one person and their unique id identifies them. However, since both datasets don't have the same number of row this means that rows that are missing in one dataset (an id that doesn't have a corresponding id in the other dataset) would have missing data values corresponding to the sections that are related to the dataset they are not in. 

In [12]:
# Check unique IDs in each dataset
print("Number of unique IDs in numerical dataset:", len(df_num['unique_id'].unique()))
print("Number of unique IDs in categorical dataset:", len(df_cat['unique_id'].unique()))

# Merge the two given numerical and categorical datasets based on their unique_ID.
df_combined = pd.merge(df_num_clean, df_cat_clean, on='unique_id')

print(df_combined)

Number of unique IDs in numerical dataset: 390
Number of unique IDs in categorical dataset: 400
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
385     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
386     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
389     926033  50.0   70.0  230.0 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [13]:
# Calculate the percentage of rows that contain at least one missing value
total_rows = len(df_combined)
total_rows2 = df_combined.shape[0]
rows_with_missing = df_combined.isnull().any(axis=1).sum()
missing_percentage = (rows_with_missing / total_rows) * 100

print("Total number of rows:", total_rows)
print("Total number of rows:", total_rows2)
print("Rows with at least one missing value:", rows_with_missing)

# Print %
print(f"Percentage of rows with missing values: {missing_percentage:.2f}%")

# Drop these rows from the dataset
# Drop rows with missing values
df_clean = df_combined.dropna()

# Print the Dataset
print("Cleaned Dataset (all rows):")
print(df_clean)

Total number of rows: 390
Total number of rows: 390
Rows with at least one missing value: 237
Percentage of rows with missing values: 60.77%
Cleaned Dataset (all rows):
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.

## Part 3.4: Sort the dataset according to the Labels

In [14]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
# Sort by Target column and reset index
df_sorted = df_clean.sort_values(by='Target').reset_index(drop=True)

# Print the sorted dataset
print("Dataset sorted by Target column:")
print(df_sorted)

Dataset sorted by Target column:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

           pc     

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [15]:
# List of categorical columns to encode
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

# Create dummy variables for categorical columns
df_encoded = pd.get_dummies(df_sorted, columns=categorical_columns)

# Print the encoded dataset
print("Dataset with encoded categorical variables:")
print(df_encoded)

# Print column names to verify encoding
print("\nNew column names:")
print(df_encoded.columns.tolist())

Dataset with encoded categorical variables:
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2       484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     ca

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: For binary variables (like 'normal'/'abnormal'), using a single column is mathematically equivalent to using dummy variables in downstream modeling. However, for categorical variables with more than 2 values (like 5 condiment types), using a single column would impose artificial ordering and equal spacing between categories, which would be incorrect. With multiple categories, using dummy variables is necessary because it allows each category to have an independent effect on the prediction without imposing relationships between them.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [16]:
# Remove outliers

# List of numerical columns to check for outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Remove outliers (values outside 3 standard deviations)
df_no_outliers = df_encoded.copy()
for col in numerical_columns:
    mean = df_encoded[col].mean()
    std = df_encoded[col].std()
    df_no_outliers = df_no_outliers[(df_no_outliers[col] >= mean - 3*std) & 
                                   (df_no_outliers[col] <= mean + 3*std)]

# Print the dataset
print(f"Rows remaining after removing outliers: {len(df_no_outliers)}")
print(df_no_outliers)
        

Rows remaining after removing outliers: 134
     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
6       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     ca

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [17]:
# Normalize the all Numerical Attributes in the dataset.
# Create a copy of the dataset
df_normalized = df_no_outliers.copy()

# Initialize the MinMaxScaler
scaler = MinMaxScaler()

# Apply normalization to numerical columns (scales values between 0 and 1)
df_normalized[numerical_columns] = scaler.fit_transform(df_normalized[numerical_columns])

# Print the dataset

# Print the dataset with normalized values
print("Dataset with normalized numerical columns:")
print(df_normalized)

# Show statistics of normalized columns to verify
print("\nStatistics of normalized numerical columns:")
print(df_normalized[numerical_columns].describe())


Dataset with normalized numerical columns:
     unique_id       age    bp       bgr        bu        sc       sod  \
0       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1       137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
3       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
4       532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
6       546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
..         ...       ...   ...       ...       ...       ...       ...   
148     647552  0.743243  0.25  0.214592  0.124183  0.037037  0.566667   
149     125587  0.418919  0.00  0.167382  0.241830  0.086420  0.700000   
150     614376  0.689189  0.50  0.210300  0.045752  0.098765  0.500000   
151     853426  0.472973  0.25  0.236052  0.183007  0.024691  0.666667   
152     572140  0.662162  0.50  0.145923  0.117647  0.061728  0.733333   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_good 

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The column which should not be used for modeling and prediction is the unique-id column since this column doesn't add any information about the patient and is arbitrarily assigned in order to be an identifier for the data. Thus it should not be included in the dataset for the modeling and predictions.

In [19]:
#Remove that column
# Remove unnecessary columns
columns_to_drop = [
    'unique_id',  # Identifier with no predictive value
]

# Remove the columns
df_final = df_normalized.drop(columns=columns_to_drop)

# Print the final dataset
print("Final dataset shape after removing unnecessary columns:", df_final.shape)
print(df_final.head())
# Print the dataset

Final dataset shape after removing unnecessary columns: (134, 35)
        age    bp       bgr        bu        sc       sod       pot      hemo  \
0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.171429  0.221311   
1  0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000  0.295082   
3  0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.657143  0.172131   
4  0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.657143  0.450820   
6  0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.314286  0.344262   

        pcv      wbcc  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0  0.184211  0.653226  ...    True    False        True       False   True   
1  0.368421  0.217742  ...   False     True       False        True   True   
3  0.210526  0.395161  ...   False     True       False        True  False   
4  0.447368  0.500000  ...    True    False        True       False   True   
6  0.315789  0.830645  ...    True    False        True  

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [26]:
# Export the dataframe to a new csv file
df_final.to_csv('kidney_disease_clean.csv', index=False)

# Export the dataframe to a new json file
df_final.to_json('kidney_disease_clean.json', orient='records')

# Confirm export
print("Data exported successfully:")
print(f"- CSV file saved as 'kidney_disease_clean.csv'")
print(f"- JSON file saved as 'kidney_disease_clean.json'")

# Print top 15 rows for part 4
print(df_final.head(15))

Data exported successfully:
- CSV file saved as 'kidney_disease_clean.csv'
- JSON file saved as 'kidney_disease_clean.json'
         age    bp       bgr        bu        sc       sod       pot  \
0   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.171429   
1   0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000   
3   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.657143   
4   0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.657143   
6   0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.314286   
8   0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.542857   
10  0.716216  0.50  1.000000  0.163399  0.111111  0.066667  0.171429   
13  0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.028571   
15  0.837838  0.50  0.373391  0.522876  0.456790  0.333333  1.000000   
17  0.000000  0.00  0.103004  0.372549  0.074074  0.500000  0.571429   
19  0.878378  0.00  0.206009  0.751634  0.604938  0.533333  0.571429   
20  0.851351

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

### Paste the markdown table here

|    |        age |   bp |       bgr |        bu |        sc |       sod |       pot |      hemo |       pcv |      wbcc | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|------------|------|-----------|-----------|-----------|-----------|-----------|------------|-----------|-----------|-----|--------|---------|------------|-------------|-------|--------|--------|---------|------------|---------------|
| 0  | 0.202703   | 0.75 | 0.158798  | 0.196078  | 0.160494  | 0.166667  | 0.171429  | 0.221311   | 0.184211  | 0.653226  | ... |   True |   False |     True   |      False   |  True |  False |  False |   True  |    True    |      False     |
| 1  | 0.905405   | 1.00 | 0.965665  | 0.522876  | 0.641975  | 0.666667  | 0.000000  | 0.295082   | 0.368421  | 0.217742  | ... |  False |    True |     False  |       True   |  True |  False |   True |  False  |    True    |      False     |
| 3  | 0.743243   | 0.50 | 0.442060  | 0.901961  | 0.432099  | 0.500000  | 0.657143  | 0.172131   | 0.210526  | 0.395161  | ... |  False |    True |     False  |       True   | False |   True |  False |   True  |    True    |      False     |
| 4  | 0.729730   | 0.75 | 0.150215  | 0.281046  | 0.234568  | 0.533333  | 0.657143  | 0.450820   | 0.447368  | 0.500000  | ... |   True |   False |     True   |      False   |  True |  False |   True |  False  |    True    |      False     |
| 6  | 0.540541   | 0.00 | 0.399142  | 0.535948  | 0.358025  | 0.700000  | 0.314286  | 0.344262   | 0.315789  | 0.830645  | ... |   True |   False |     True   |      False   |  True |  False |   True |  False  |    True    |      False     |
| 8  | 0.675676   | 0.75 | 0.253219  | 0.633987  | 0.777778  | 0.366667  | 0.542857  | 0.286885   | 0.342105  | 0.169355  | ... |   True |   False |     True   |      False   |  True |  False |   True |  False  |    True    |      False     |
| 10 | 0.716216   | 0.50 | 1.000000  | 0.163399  | 0.111111  | 0.066667  | 0.171429  | 0.393443   | 0.500000  | 0.532258  | ... |   True |   False |     False  |      True    |  True |  False |   True |  False  |    True    |      False     |
| 13 | 0.729730   | 0.00 | 0.935622  | 0.169935  | 0.160494  | 0.333333  | 0.028571  | 0.188525   | 0.236842  | 0.879032  | ... |   True |   False |     False  |      True    |  True |  False |  False |   True  |    True    |      False     |
| 15 | 0.837838   | 0.50 | 0.373391  | 0.522876  | 0.456790  | 0.333333  | 1.000000  | 0.000000   | 0.000000  | 0.540323  | ... |  False |    True |     False  |      True    |  True |  False |   True |  False  |    True    |      False     |
| 17 | 0.000000   | 0.00 | 0.103004  | 0.372549  | 0.074074  | 0.500000  | 0.571429  | 0.352459   | 0.368421  | 1.000000  | ... |   True |   False |     False  |      True    |  True |  False |   True |  False  |    True    |      False     |
| 19 | 0.878378   | 0.00 | 0.206009  | 0.751634  | 0.604938  | 0.533333  | 0.571429  | 0.475410   | 0.500000  | 0.879032  | ... |   True |   False |     False  |      True    |  True |  False |   True |  False  |    True    |      False     |
| 20 | 0.851351   | 0.25 | 0.618026  | 0.562092  | 0.728395  | 0.000000  | 0.285714  | 0.311475   | 0.315789  | 0.580645  | ... |  False |    True |     True   |      False   |  True |  False |   True |  False  |    True    |      False     |
| 21 | 0.878378   | 0.25 | 0.639485  | 0.470588  | 0.395062  | 0.428571  | 0.428571  | 0.393443   | 0.447368  | 0.104839  | ... |  False |    True |     True   |      False   |  True |  False |   True |  False  |    True    |      False     |
| 22 | 0.783784   | 0.00 | 0.725322  | 0.313725  | 0.481481  | 0.714286  | 0.714286  | 0.319672   | 0.342105  | 0.258065  | ... |   True |   False |     False  |      True    |  True |  False |   True |  False  |    True    |      False     |
| 24 | 0.662162   | 0.50 | 0.618026  | 0.411765  | 0.432099  | 0.571429  | 0.571429  | 0.434426   | 0.473684  | 0.250000  | ... |   True |   False |     True   |      False   |  True |  False |   True |  False  |    True    |      False     |

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [24]:
# Rename mapping based on the provided information
rename_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wc': 'White Blood Cell Count',
    'rc': 'Red Blood Cell Count',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad': 'Coronary Artery Disease',
    'appet': 'Appetite',
    'pe': 'Pedal Edema',
    'ane': 'Anemia',
    'class': 'Class'
}

# Apply the renaming to your dataframe
df_renamed = df_normalized.rename(columns=rename_mapping)

# Print the dataframe with new column names
print("Dataset with descriptive column names:")
print(df_renamed.head(15))

Dataset with descriptive column names:
    unique_id       Age  Blood Pressure  Blood Glucose Random  Blood Urea  \
0      474407  0.202703            0.75              0.158798    0.196078   
1      137148  0.905405            1.00              0.965665    0.522876   
3      343710  0.743243            0.50              0.442060    0.901961   
4      532520  0.729730            0.75              0.150215    0.281046   
6      546225  0.540541            0.00              0.399142    0.535948   
8      514721  0.675676            0.75              0.253219    0.633987   
10     621332  0.716216            0.50              1.000000    0.163399   
13     481293  0.729730            0.00              0.935622    0.169935   
15     297893  0.837838            0.50              0.373391    0.522876   
17     615697  0.000000            0.00              0.103004    0.372549   
19     843362  0.878378            0.00              0.206009    0.751634   
20     995177  0.851351            0.

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [25]:
# Import pandas (if not already imported)
import pandas as pd

# Load both datasets (if not already loaded)
df_num = pd.read_csv('chronic_kidney_disease_numerical.csv')
df_cat = pd.read_csv('chronic_kidney_disease_categorical.csv')

# Combine datasets (assuming they have matching indices)
df_combined = pd.merge(df_num, df_cat, on='unique_id', how='outer')

# Drop rows with missing values
df_clean = df_combined.dropna()

# Now run the groupby operation on the cleaned dataset
result = df_clean.groupby('Target').size().reset_index(name='count')

# Print the result
print(result)

# Converted SQL to Pandas code
import pandas as pd

# Assuming your DataFrame is named df
result = result.groupby('Target').size().reset_index(name='count')

   Target  count
0     ckd     48
1  notckd    150


*Describe what the above code does here*

The code splits the data into two categories, ckd and notckd.

There are 48 patients with chronic kidney disease (ckd)
There are 150 patients without chronic kidney disease (notckd)

This reveals an imbalanced dataset with about 3 times more non-kidney disease cases than kidney disease cases. This information is important for model training, as it suggests strategies like class weighting might be needed to prevent bias toward the majority class in machine learning models.